modin/backends/pandas/query_compiler.py

```diff
@@ -561,21 +561,92 @@ def reset_index(self, **kwargs):
         """
         drop = kwargs.get("drop", False)
         level = kwargs.get("level", None)
-        # TODO Implement level
-        if level is not None or self.has_multiindex():
-            return self.default_to_pandas(pandas.DataFrame.reset_index, **kwargs)
+        new_index = None
+        if self.has_multiindex() and level is not None:
+            if not isinstance(level, (tuple, list)):
+                level = [level]
+            level = [self.index._get_level_number(lev) for lev in level]
+            uniq_sorted_level = sorted(set(level))
+            if len(level) < self.index.nlevels:
+                new_index = self.index.droplevel(uniq_sorted_level)
         if not drop:
-            new_column_name = (
-                self.index.name
-                if self.index.name is not None
-                else "index"
-                if "index" not in self.columns
-                else "level_0"
-            )
-            new_self = self.insert(0, new_column_name, self.index)
+            # Some parts of index remain in place
+            if self.has_multiindex():
+                col_level = kwargs.get("col_level", 0)
+                col_fill = kwargs.get("col_fill", "")
+
+                # Convert multiindex into a dataframe with columns containing multiindex levels
+                mi_frame = self.index.to_frame()
+                # Overwrite constructed columns index because it ruins tuples with NaNs since
+                # tupleize_cols is True by default
+                mi_frame.columns = pandas.Index(self.index.names, tupleize_cols=False)
+
+                if level is not None:
+                    mi_frame = mi_frame.iloc[:, uniq_sorted_level]
+                if isinstance(self.columns, pandas.MultiIndex):
+                    items = []
+                    if col_fill is None:
+                        # Initialize col_fill if it is None.
+                        # This is some weird undocumented Pandas behavior to take first
```

Comment:
Maybe instead trying to emulate pandas behaviour in terms of constructing new index and columns, we should try to create an empty pandas frame with the modins index and columns, apply `reset_index` on it with all passed parameters and then extract proper `new_index` and `new_columns` 
```python
dummy_df = pandas.DataFrame(index=md_df.index[:1], columns=md_df.columns).reset_index(*args, **kwargs)
new_index, new_columns = dummy_df.index, dummy_df.columns
```
but we should bench this approach and compare its time with the current implementation

Reply:
This will likely depend on the parameter combinations and we currently don't have any `reset_index` benchmarks, not to mention with multiple different parameter combinations. BTW why do you want to slice index as `[:1]`? Won't it make reset index concatenate only 1st row from the index levels into the frame?

Reply:
We need it to get new proper columns, while building new columns `reset_index` considers only old level names, which is not dependent on the index length. We can even take 0 elements from the original index since we need only meta information without values.

<details><summary> Code example </summary>


```python
import numpy as np
import modin.pandas as pd
import pandas
from modin.pandas.test.utils import df_equals, generate_multiindex, create_test_dfs

index = generate_multiindex(5)
columns = generate_multiindex(5)
data = {col: np.arange(5) for col in columns}

md_df, pd_df = create_test_dfs(data, index=index, columns=columns)
dummy_df = pandas.DataFrame(index=md_df.index[:0], columns=md_df.columns)

md_res, pd_res = dummy_df.reset_index(drop=False), pd_df.reset_index(drop=False)
print(md_res.columns.equals(pd_res.columns))  # True
```



</details>

Reply:
I've tried this approach on top of the copy of this branch in [this commit](https://github.com/dchigarev/modin/commit/ab0fb15e84a36b5d68775e97f7a432cc8674d2a0) and all of the tests passed.

Reply:
Yes I understand that you should get correct columns with `md_df.index[:1]` but my question was about assigning `dummy_df.index` to `new_index` in 2nd line of your code. You are supposed to concatenate all index rows into the frame columns.

Reply:
Yes, you're right. We can't retrieve `new_index` from `dummy_df`.